Adding accessible markup to sale price.

BFTrick opened the pull request:
### Changes proposed in this Pull Request:

This PR adds accessible markup to the sale price.

A user who only uses screen reader software can now hear and understand the original price as well as the current (sale) price.

This is important for accessibility.

Here's a demo (🔉 on):

https://github.com/woocommerce/woocommerce/assets/1065372/e35fb5af-aac0-4af0-85c6-64d74d3e2f87

Full issue: https://github.com/woocommerce/woocommerce/issues/31099

### How to test the changes in this Pull Request:
1. Create a simple product in WC > Products with a sale price set.
2. Click its permalink on the admin or search for the product on the frontend and go to the product's page.
3. Using the devtools inspector choose the area around the product price (or look for the `<p class="price">` tag).
4. Confirm that both the `<del>` and `<ins>` tags are present and have `aria-hidden` set. These tags should include the original and sale price, respectively (see screenshot below).
5. Confirm that there are a couple of `<span>` tags with class `screen-reader-text` which clarify with text what the original and current prices are (see screenshot below).

<img width="465" alt="Screenshot 2024-03-14 at 09 28 29" src="https://github.com/woocommerce/woocommerce/assets/184724/8204638b-0d58-403c-84ae-a784d9f58e75">

BFTrick committed:
Adding accessible markup to sale price. See #31099

This PR adds accessible markup to the sale price.

A user who only uses screen reader software can now hear and understand the original price as well as the current price.

This is important for accessibility.

Full issue: https://github.com/woocommerce/woocommerce/issues/31099

jorgeatorres reviewed (changes requested):
Hi @BFTrick!

Thanks for your contribution. This is a much appreciated change :100.
Thinking of RTL languages, I believe ideally the translatable strings inside `screen-reader-text` should contain a placeholder for the actual value. For example:

```php
if ( is_numeric( $regular_price ) ) {
	$price .= sprintf( esc_html__( 'Original price was: %s.', 'woocommerce' ), wc_price( $regular_price ) );
}
```

Otherwise, the text (once translated) along with the price might not make the most sense for certain languages.

Also wanted to note that the new approach isn't equivalent to the current one. When either `$regular_price` or `$sale_price` are not numeric we currently don't output anything between the `<del>` or `<ins>` tags, but with this new approach we will output, for example, "Current price is:" when that variable is empty.
Ideally we wouldn't add unnecessary output, so we might need to adjust the checks here.

Would it make sense to leave the `<del>` tag alone with `aria-hidden=true` and add an `<span class="screen-reader-text">` with "Original price was: X" immediately after it and before the sale price? With that in mind, we might not need to clarify what the "current" price is for screen readers.

Please let me know what you think and, once again, thanks for your contribution!

BFTrick committed:
Improving sale price string formatting & localization #31099

This PR merges in feedback from the WooCommerce team.

* Improving string translation & output. Giving RTL languages more customization with the price in the translatable string.
* Formatting regular price & sale price at the top and then inserting into the HTML below. This prevents some sloppy logic where you might print out a blank del or ins tag.

BFTrick commented:
Thanks for the thoughtful feedback @jorgeatorres 🙏 

> Thinking of RTL languages, I believe ideally the translatable strings inside `screen-reader-text` should contain a placeholder for the actual value

Agreed. Much better solution!

✅  Updated code with this suggestion 

> When either `$regular_price` or `$sale_price` are not numeric we currently don't output anything between the `<del>` or `<ins>` tags

Derp! Definitely should avoid printing empty tags. 👍 

✅ I updated my code so I do my checks and format the price at the top of the function then use that value lower down. 

> Would it make sense to leave the `<del>` tag alone with `aria-hidden=true` and add an `<span class="screen-reader-text">` with "Original price was: X" immediately after it and before the sale price? 

From an accessibility point of view either will work. As long as the information is conveyed it's accessible. 👍 

With localization in mind: `$price .= sprintf( esc_html__( 'Original price was: %s.', 'woocommerce' ), $formatted_regular_price );` it's easier & more straightforward to format with a `<del>` tag that's hidden and a separate `<span class="screen-reader-text">` immediately after.

✅ I updated my code with this.

> With that in mind, we might not need to clarify what the "current" price is for screen readers.

I'll disagree here. We should be explicit with the current price. Otherwise it sounds like:

> Original price was $XX. $XX.

There's no context for the second number.

Here's without clarification (🔉  on):

https://github.com/woocommerce/woocommerce/assets/1065372/8f67e7de-2af5-4eb4-b6e5-a2f1bccec745

Here's with clarification (🔉  on):

https://github.com/woocommerce/woocommerce/assets/1065372/3ed87b42-fe13-40fa-a841-d1d686805b9a

jorgeatorres commented on plugins/woocommerce/includes/wc-formatting-functions.php:
```diff
@@ -1288,7 +1288,27 @@ function wc_format_stock_quantity_for_display( $stock_quantity, $product ) {
+
+	// for accessibility (a11y) we'll also display that information to screen readers
+	$price .= '<span class="screen-reader-text">';
+	$price .= sprintf( esc_html__( 'Original price was: %s.', 'woocommerce' ), $formatted_regular_price );
```
We should probably strip HTML tags in `$formatted_regular_price` so that the end result in the `screen-reader-text` span is just plain text.

This is one way to do it, for example:

```suggestion
	$price .= esc_html( sprintf( __( 'Original price was: %s.', 'woocommerce' ), wp_strip_all_tags( $formatted_regular_price ) ) );
```

Notice that I've reversed the `esc_html()` and `sprintf()`. This is intentional because if translations accidentally include some HTML, that could break things.

- - -

With this, the end result will be

<img width="537" alt="Screenshot 2024-02-22 at 18 43 48" src="https://github.com/woocommerce/woocommerce/assets/184724/5d3e0143-698c-40e6-98d9-9e9a9e767e7b">

instead of

<img width="464" alt="Screenshot 2024-02-22 at 19 03 12" src="https://github.com/woocommerce/woocommerce/assets/184724/62379ed6-f742-42b2-a08f-2cf996893073">

BFTrick replied:
Great idea! Appreciate your thoroughness!

Added in the commit below.

jorgeatorres commented on plugins/woocommerce/includes/wc-formatting-functions.php:
```diff
@@ -1288,7 +1288,27 @@ function wc_format_stock_quantity_for_display( $stock_quantity, $product ) {
+
+	// for accessibility (a11y) we'll also display that information to screen readers
+	$price .= '<span class="screen-reader-text">';
+	$price .= sprintf( esc_html__( 'Current price is: %s.', 'woocommerce' ), $formatted_sale_price );
```
Same as above but for `$formatted_sale_price` instead.

BFTrick replied:
Great idea! Appreciate your thoroughness!

Added in the commit below.

jorgeatorres reviewed (changes requested):
Hi @BFTrick!

Thanks for making those adjustments. I agree with your decisions here. The code looks good to me, but there are just a few minor things to address.

- There are a few coding standards violations in the code (basically spacing related), which you can see here: https://github.com/woocommerce/woocommerce/actions/runs/7944951176/job/21691017251.
- You have to add a changelog file to the PR. You can do that by running command `pnpm --filter=@woocommerce/plugin-woocommerce changelog add` and responding to the questions in the wizard. Let me know if you're having trouble with that.
- The changes in the HTML broke some of the E2E tests. If you're not sure how to address those, I'll take a look myself.
- A few suggestions (which I've left as comments in the code review).

Once again, thanks for working on this and for your patience during review. Much appreciated.

BFTrick committed:
Swapping esc_html & sprintf #31099

Reversed the esc_html() and sprintf(). This is intentional because if translations accidentally include some HTML, that could break things.

Also, stripping HTML tags in $formatted_regular_price so that the end result in the screen-reader-text span is just plain text.

BFTrick committed:
Adjusting white space #31099

Mostly removing white space for the lint checker

BFTrick committed:
Adding translator comments #31099

Adding translator comments for lint

BFTrick commented:
@jorgeatorres

Appreciate your feedback! I've been trying to work on this for 2 weeks now but just haven't had to the time. We're launching a big sale in just a couple days so that's taken priority.

I merged in your feedback. I did my best to fix the whitespace issues from Lint but I'm not sure where all of them are coming from. The line numbers seem off. 🤔 

![Screenshot 2024-03-11 at 4 53 09 PM](https://github.com/woocommerce/woocommerce/assets/1065372/402732a8-4810-44ac-8b72-927bbc72de5c)

I generally cleaned up the whitespace though. 🧹 

I also don't have my repo setup the right way to build the changelog 🙈 

Is it possible for you to carry this across the finish line? 🤞

jorgeatorres committed:
Add changelog

jorgeatorres committed:
Fix some PHPCS violations

jorgeatorres committed:
Fix unit tests

jorgeatorres committed:
Adjust E2E test

jorgeatorres reviewed (approved):
LGTM. Thank you for your work on this @BFTrick!

I've added the changelog file and addressed a few issues with the unit and E2E tests due to the change in markup. We should be good now.
